### PR TITLE
feat: Consult global cluster for namespace creation

### PR DESCRIPTION
**Pull request by JigarJoshi**
## Describe your changes
In multiregion deployment, there will be a global cluster responsible for namespace creation and rest of the clusters will consult that global cluster in case if they don't find it locally.

### CODE REVIEW

File: server/services/v1/database/sessions.go [hunk 1]
```diff
@@ -395,6 +414,54 @@ func (s *QuerySession) Commit(versionMgr *metadata.VersionHandler, incVersion bo
 	return err
 }
 
+func (sessMgr *SessionManager) consultCentralClusterForNamespace(ctx context.Context, namespaceId string) (*metadata.Tenant, error) {
```
Comment (himank) [inline server/services/v1/database/sessions.go:417]: Should we call it `askMetadataCluster`?

File: server/config/options.go [hunk 2]
```diff
@@ -38,6 +38,7 @@ type Config struct {
 	Log            log.LogConfig
 	Server         ServerConfig         `yaml:"server" json:"server"`
 	Auth           AuthConfig           `yaml:"auth" json:"auth"`
+	CentralCluster ClusterConfig        `mapstructure:"central_cluster" yaml:"central_cluster" json:"central_cluster"`
```
Comment (himank) [inline server/config/options.go:41]: `MetadataCluster` Or `UserCluster`

Comment (ovaistariq) [reply server/config/options.go:41]: `MetadataCluster` would be better

File: server/services/v1/database/sessions.go (hunk 1 above)
Comment (JigarJoshi) [reply server/services/v1/database/sessions.go]: renamed.

[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/database/sessions.go]: renamed to metadata cluster.

File: server/services/v1/database/sessions.go [hunk 3]
```diff
@@ -395,6 +414,54 @@ func (s *QuerySession) Commit(versionMgr *metadata.VersionHandler, incVersion bo
...
+		return nil, err
+	}
+
+	if listNamespaceResp.GetNamespaces() == nil || len(listNamespaceResp.GetNamespaces()) == 0 {
+		log.Error().Err(err).Str(Component, NamespaceConsultation).Msg("Failed to find namespace")
+		return nil, err
+	}
+
+	namespaceToCreate := listNamespaceResp.Namespaces[0]
+	return sessMgr.tenantMgr.CreateOrGetTenant(ctx, metadata.NewTenantNamespace(namespaceToCreate.GetId(), metadata.NamespaceMetadata{
```
[the remark this answers is not in the document]
Comment (JigarJoshi) [reply server/services/v1/database/sessions.go]: updated.

File: server/services/v1/database/sessions.go [hunk 4]
```diff
@@ -33,6 +38,11 @@ import (
 	ulog "github.com/tigrisdata/tigris/util/log"
 )
 
+const (
+	NamespaceConsultation = "namespace_consultation"
```
Comment (efirs) [inline server/services/v1/database/sessions.go:42]: Would "federation" be a better term here?
WDYT @JigarJoshi @himank @ovaistariq

Comment (ovaistariq) [reply server/services/v1/database/sessions.go:42]: Here we are checking from a central place so probably not federation.

Comment (himank) [reply server/services/v1/database/sessions.go:42]: `NamespaceConsultation` or `MetadataClusterConsultation` sounds good to me.

Comment (efirs) [reply server/services/v1/database/sessions.go:42]: ok.

Comment (JigarJoshi) [reply server/services/v1/database/sessions.go]: Discussed offline and decided to rename it to `NamespaceLocalization`

File: server/services/v1/database/sessions.go [hunk 5]
```diff
@@ -180,8 +190,27 @@ func (sessMgr *SessionManager) Create(ctx context.Context, trackVerInOwnTxn bool
 	}
 	tenant, err := sessMgr.tenantMgr.GetTenant(ctx, namespaceForThisSession)
 	if err != nil {
-		log.Warn().Err(err).Msgf("Could not find tenant, this must not happen with right authn/authz configured")
-		return nil, errors.NotFound("Tenant %s not found", namespaceForThisSession)
+		// localize the namespace from metadata cluster.
+		if config.DefaultConfig.Auth.NamespaceLocalization.Enabled {
```
Comment (himank) [inline server/services/v1/database/sessions.go:194]: I thought if this flag is true then it is localized i.e. it won't consult the metadata cluster.

**Review by himank — APPROVED** [review]

File: server/services/v1/database/sessions.go (hunk 5 above)
Comment (JigarJoshi) [reply server/services/v1/database/sessions.go:194]: if localization is enabled - then it will get it from metadata cluster and apply in local cluster.

**Pull request merged**